omp/components/sort_small.hpp

```diff
@@ -0,0 +1,200 @@
+/*******************************<GINKGO LICENSE>******************************
+Copyright (c) 2017-2020, the Ginkgo authors
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_OMP_COMPONENTS_SORT_SMALL_HPP_
+#define GKO_OMP_COMPONENTS_SORT_SMALL_HPP_
+
+
+#include <algorithm>
+
+
+#include <ginkgo/core/base/types.hpp>
+
+
+namespace gko {
+namespace kernels {
+namespace omp {
+
+
+#ifdef __GNUG__
+#define GKO_GCC_EXPECT_LIKELY(v) __builtin_expect(v, true)
+#else
+#define GKO_GCC_EXPECT_LIKELY(v) v
+#endif
+
+#ifdef _MSC_VER
+#define GKO_COMPARATOR(x, y)                            \
+    {                                                   \
+        const auto should_swap = !comp(arr[x], arr[y]); \
+        const auto tmp = arr[x];                        \
+        arr[x] = should_swap ? arr[y] : arr[x];         \
+        arr[y] = should_swap ? tmp : arr[y];            \
+    }
+#else
+#define GKO_COMPARATOR(x, y)                                         \
```

Comment:
I didn't come up with the name myself, it's just convention to call the swap-if-cond things in sorting networks comparators, sorting networks are sometimes also called comparator networks.

I can actually make the comparator a function, I checked with my benchmarks, that should be fine. That would then mean parameterizing over `comp` and `nodes` too. Only thing still missing is a cross-plattform `FORCEINLINE` macro - we have `GKO_INLINE`, but it seems to only resolve to `inline`, not `__attribute__((always_inline)) inline`. Is this intended, should there be a separate, `GKO_FORCEINLINE` macro?

Reply:
Random name suggestion: `GKO_SORT2`

Reply:
@lksriemer Does a simple `inline` not work in this context?
The reason for not using `__attribute__((always_inline))` is simply because it is not portable. It is not defined by C++, therefore, we should not use or rely on the support of it (Microsoft Visual Studio does not support it for example).
An alternative would be `[[always_inline]]`, but only C++17 specified that unknown Attributes must not cause a compiler error. Since we currently use C++14, it is possible that it does not compile on some compilers.

Here, I don't see a reason why the compiler would not inline it with the keyword `inline`. It might not do it in a debug build, but that should not be a problem.

Reply:
Works fine with this code. In earlier iterations, when larger sizes were explicitly matched, compilers were reluctant to inline.
I had to introduce an extra closure (called it `sort2`) to cut down on the character count.